ENH: Add new scipy.fft module using pocketfft

peterbell10 opened the pull request:
Closes #10175

## Enhancements
* Support for long double transforms
* ~Multi-threading using OpenMP~ (disabled for now)
* NumPy-like real to complex `rfft`s including n-dimensional transforms `rfftn`
* Orthonormal transforms (`norm = 'ortho'`)
* Bluestein's algorithm
* GIL is released during transforms

## Issues
- [x] The build definitions don't select the `extra_compile_args` based on the compiler which would be needed to compile with OpenMP in a portable way.
- [x] FFTPACK optimized for the case when real sequences were passed to `fft` which is missing from pocketfft currently.
- [x] Normalization factors are only passed to c++ with double precision, even for long double transforms.
- [x] This adds a build time dependency on pybind11 and I'm not sure if it is required at run-time as well.
- [x] Docstrings need to be written
- [x] The pocketfft backend is passing all of the `numpy.fft` tests and _most_ of the `scipy.fftpack` tests after some modifications. I have 1 remaining failure locally which is a precision difference for float32 transforms. Needs investigation.

## Open questions
- [ ] How portable is OpenMP? (#10239)
- [ ] How should it be reported if multi-threading isn't available. @rgommers suggested we should `raise` when the user specifies a number of threads that isn't available. (#10242)
- [ ] NumPy's hermitian transforms (`hfft`, `ihfft`) are missing. This is trivial to add since it is only a slight variant on `rfft` and `irfft`. We could also extend beyond numpy and give n-dimensional versions `hfftn` and `ihfftn`. (#10240)
- [x] Some people find the argument `overwrite_x` to be confusing (cupy/cupy#1988). It has been suggested to remove it entirely but could also be renamed. FFTW calls this `FFTW_DESTROY_INPUT` which makes it clearer that this is __not__ a guarunteed inplace transform/out-parameter. (#10241)

rgommers commented:
Thanks @peterbell10, great start!

I created a new `scipy.fft` label for the new submodule.

For all the things that need significant discussion and deciding, I suggest to create separate issues. And keep the list in this PR as the "master issue". Otherwise things will get unwieldy. Suggestion of separate issues:
- OpenMP portability and build config (incl `pybind11`).
- what functions to include in the new module (e.g. `hfft`)
- multithreading behavior
- `overwrite_x`

peterbell10 commented:
Thanks @rgommers, that's a good idea.

peterbell10 commented:
Regarding the failing test, single precision fftpack tests were marked as xfail so fftpack failed for the same test case. Investigating it shows that pocketfft is in fact more accurate than fftpack.

```
Max abs diff fft:      2.99796178049074e-06
Max abs diff fftpack:  3.06302563664315e-06
Max rel diff fft:      4.73642558242986e-05
Max rel diff fftpack:  2.63469637785951e-04
```

I've given single precision variants slightly more tolerance so the test should be passing now.

peterbell10 commented:
I've had a look through the [pybind11 repo][1] and the only python code is either tests or the `get_include` functionality. That, combined with the fact that there is nothing else linked in; I'm now reasonably confident that it's only a build-time dependency.

[1]: https://github.com/pybind/pybind11

peterbell10 commented:
Also, I'm planning to base a lot of the documentation off the numpy fft docs as they seemed to be slightly more consistent. I hope that's not a problem.

rgommers commented:
> Also, I'm planning to base a lot of the documentation off the numpy fft docs as they seemed to be slightly more consistent. I hope that's not a problem.

peterbell10 commented:
I'm struggling to get the documentation to build correctly. The `scipy.fft` module docstring gives me a numpydoc error, despite it being very similar to every other module docstring 

    Unknown section Fast Fourier Transforms (ffts) in the docstring of <module 'scipy.fft'[...]
which comes from this section heading in the docstring, which is exactly the same as for fftpack
    
    Fast Fourier Transforms (FFTs)
    ==============================

If I remove the section headings entirely, the module summary builds. However, there are no `.rst` files for the `fft` module in `doc/source/generated` so the functions aren't documented anywhere.  I assume I need to tell the generator about the fft module somewhere but I can't see where the generator is.

Any help would be appreciated.

rgommers commented on scipy/fft/__init__.py:
```diff
@@ -0,0 +1,90 @@
+"""
+==================================================
+Discrete Fourier transforms (:mod:`scipy.fft`)
+==================================================
```
under/over-line length are recommended to be the same. longer is acceptable so I don't think it's your build problem, but same length is better

rgommers commented:
> `Unknown section `

Hmm, that error is normally for function docstrings. Perhaps `numpydoc` gets itself confused and is parsing a module docstring as function docstring. Could be because `scipy.fft` used to be a function and you didn't do a clean rebuild?

peterbell10 commented:
CircleCI should be doing a clean build and it's happening on there as well.

peterbell10 commented:
It looks like the root cause is that `isinstance(scipy.fft, Callable)==True`, so numpydoc is treating it like a function.

peterbell10 commented:
I've added a workaround for sphinx documenting `scipy.fft` as a function and the numpydoc error is gone. However, the function documentation still isn't begin generated.

I've also been trying to update the `setup.py` to conditionally add `--std=c++11` for gcc-like compilers.  The correct way to do this seems to be providing a [custom `build_ext` class][1] but there doesn't seem to be a standard way to update `cmdclass` in the `numpy.distutils` `Configuration` object.

[1]: https://github.com/pybind/python_example/blob/master/setup.py

grlee77 commented on azure-pipelines.yml:
```diff
@@ -23,7 +23,7 @@ jobs:
            python3.6 get-pip.py && \
            pip3 --version && \
-           pip3 install setuptools wheel numpy cython==0.29.2 pytest pytest-timeout pytest-xdist pytest-env pytest-faulthandler pytest-cov Pillow mpmath matplotlib --user && \
+           pip3 install setuptools wheel numpy cython==0.29.2 pybind11 pybind11 pytest pytest-timeout pytest-xdist pytest-env pytest-faulthandler pytest-cov Pillow mpmath matplotlib --user && \
```
pybind11 appears here twice?

peterbell10 commented:
I've fixed the build issues by writing a `_pre_build_hook` function into the `Extension` object that gets called before building. That way I can add the compile arguments while knowing what compiler is actually going to be used.

I've also fixed importing individual files from `fftpack` like `scipy.fftpack.helper`. I'm not sure if this is meant to be supported, but the examples for `next_fast_len` import it this way, so I presume people will be using it. Unfortunately, the diff doesn't seem to follow the renamed files after they've been replaced which is inflating the diff. The history is still preserved though and `git blame` still works.

With that, the issues list has all been addressed and I'm just waiting for reviews. (@grlee77 @larsoner ?)

rgommers commented on scipy/fft/_pocketfft/setup.py:
```diff
@@ -0,0 +1,74 @@
+    if compiler.compiler_type == 'msvc':
+        flags = ['/std:c++17', '/std:c++14', None]
+    else:
+        flags = ['--std=c++17', '--std=c++14', '--std=c++11']
```
The Intel compiler flag is platform dependent IIRC, uses MSVC-like `/xxx` on Windows and regular (`-`) flags on other platforms. See note under the table in https://software.intel.com/en-us/articles/c17-features-supported-by-intel-c-compiler and `numpy/distutils/intelccompiler/py`.

peterbell10 replied:
`intelw` added. I can't test it but the [docs](https://software.intel.com/en-us/cpp-compiler-developer-guide-and-reference-std-qstd) say the form is `/Qstd=c++##` on windows.

rgommers commented:
Awesome. It would be good to give a short summary on the mailing list as well and ask for inputs/review. We don't add a new submodule very often. Actually, it's the first `scipy.<newmodule>` one in a decade I think:)

Regarding merge strategy and potential blockers:
- the "`scipy.fft` is now a function thing" needs a decision
- the C++14 and OpenMP portability issue needs testing, but perhaps it's best to be aggressive in merging here so it's easier to test on exotic platform/compilers - we can ask users for help here. There's 4-5 months left till the next release, so this is a good time for merging.
- no other blockers I can think of.

peterbell10 commented:
A couple notes:
- only C++11 is a hard requirement
- No OpenMP code is compiled here

rgommers commented:
> My vote is to live with this hack for a suitable deprecation period (2 cycles?) during which we also deprecate the other from numpy import * functions. After that we can get rid of it. This seems like the best long-term solution.

Agreed. I played around with various imports and ways of calling `fft`, and it works well enough. There are things like `fft?` for which the behavior changes, but that's acceptable I'd think.

rgommers commented:
I had a look at the docs, those look good. A couple of small things to do:
- in the tutorial, replace `fftpack` with `fft` and remove the last small section about fftpack cache
- at the top of the `fftpack` refguide section, add a warning box that `scipy.fft` should be used for new code and that `fftpack` is legacy.
- In the main TOC, change the fftpack entry name from "Discrete Fourier Transforms" to "Legacy Discrete Fourier Transforms"

peterbell10 commented:
I've updated the docs as you asked @rgommers and also wrote up the conventions for `rfft` and `irfft`.

rgommers commented:
This is starting to look close to merge-ready.

One thing I didn't see yet is where it's documented what exact version of pypocketfft is vendored. Perhaps good to add to the included README?

peterbell10 commented:
I don't think there is any version numbering for pypocketfft, would a commit hash suffice? I've also made a few modifications to pypocketfft.cxx, should I also mention that in the README?

rgommers commented:
Yes, commit hash would be great. 

>  I've also made a few modifications to pypocketfft.cxx, should I also mention that in the README?

Yes please, that would be very helpful when updating. Or actually, better than modifying the README would be a separate file like `SCIPY_CHANGES` or something. That way a new pypocketfft version can be easily dropped in without overwriting the README changes.

Here is an example of keeping track of the changes needed for integration into SciPy so they can be easily re-applied when updating the vendored library: https://github.com/scipy/scipy/blob/master/scipy/sparse/linalg/dsolve/SuperLU/scipychanges.patch

peterbell10 commented:
I wasn't confident which commit hash I'd based my changes off of. So, I've manually "rebased" my changes onto the latest commit. I've also seperated as much as possible out of the `pypocketfft.cxx` source file to make future patching easier.

As an added benefit, this version if pypocketfft now releases the GIL while doing the transforms.

grlee77 commented on scipy/fft/_pocketfft/pypocketfft.cxx:
```diff
@@ -174,22 +188,23 @@ template<typename T> py::array irfftn_internal(const py::array &in,
   {
   py::gil_scoped_release release;
-  c2r(dims_out, s_in, s_out, axes, d_in, d_out, T(fct), nthreads);
+  c2r(dims_out, s_in, s_out, axes, d_in, d_out, fct, nthreads);
```
can't the call to `norm_fct` also go within the section with the GIL released? (the same pattern with `norm_fct` occurs elsewhere as well)

peterbell10 replied:
I've moved them all to a scoped release region.

larsoner commented:
@peterbell10 we have been trying lately to add at least some basic benchmarks with new enhancements. Since `pocketfft` should be (much, in some cases) faster than `fftpack`, it seems like this PR could benefit from adding some `asv` benchmarks of the new `scipy.fft` module. A good start might just be just doing what is already done for fftpack:

https://github.com/scipy/scipy/blob/master/benchmarks/benchmarks/fftpack_basic.py

Maybe

1. Rename this file to `fft_basic.py`
2. Add params for all tests to test both `scipy.fft` and `scipy.fftpack` versions

What do you think?

larsoner commented on scipy/__init__.py:
```diff
@@ -85,6 +85,13 @@
 
+# The existence of the scipy.fft function breaks doc builds
+import sys
+if 'sphinx' in sys.modules:
```
This is not the most robust check. A slightly more robust one would be to do `os.environ['_SCIPY_BUILDING_DOC'] = 'true'` in `doc/source/conf.py` (before any `scipy` import) and then here `if os.getenv('_SCIPY_BUILDING_DOC', '') == 'true'` or so.

larsoner commented on scipy/fft/_fftpack/__init__.py:
```diff
@@ -0,0 +1,109 @@
+"""
+==================================================
+Discrete Fourier transforms (:mod:`scipy.fftpack`)
+==================================================
```
This is a `_fftpack` module you do not need all of this docstring.

Can you remind me why it helps to move all the files from `scipy/fftpack` to `scipy/fft/_fftpack/`? It does not seem like it should be necessary. Even thinking of the upcoming backend stuff it seems like it should not matter where these (legacy) fftpack backend files live.

peterbell10 replied:
Sorry, I missed the second part of your comment here. I mentioned the reason before in #10222.

If the intention is to move `scipy.fftpack` over to using pocketfft as well, then it creates a circular dependency issue.  `fft` depends on `fftpack` for the cosine and sine transforms; `fftpack` would depend on `fft` for the normal FFTs.

In the current state (where `fftpack` doesn't use pocketfft) it's not necessary and I could revert it.

peterbell10 commented:
I'll take a look at the benchamrks now.

peterbell10 commented:
Running the benchmarks, it looks like there is a performance regression for small 1d ffts. Every call is ~50 us slower. I think this is because of the overhead of `_init_nd_shape_and_axes` but will investigate more.

mreineck commented:
> I don't think there is any version numbering for pypocketfft

That's absolutely correct ... I was still waiting for feedback before making any "official" release.
If you can live with a commit hash, I'm happy with that, but I can of course also add a tag.

Great that you updated to the version releasing the GIL!

peterbell10 commented:
I had hoped that I wouldn't have to have almost identical functions for 1d and nd transforms but I wasn't able to bring down the overhead of `_init_nd_shape_and_axes` enought that it was reasonable.

larsoner commented:
It's okay, there is often a trade-off between DRY and performance. Can you post your results from benchmarking? Hopefully it shows that `scipy.fft` wins over `scipy.fftpack`. And maybe also add `np.fft` to the benchmark suite, too, for comparison purposes?

peterbell10 commented:
@larsoner is it possible to get graphs from the output of `runtests.py --bench`? I think I can get them from `--bench-compare` but it isn't building in parallel so is taking forever.

larsoner commented:
I don't know, I usually just look at the outputs printed to the console to check. @tylerjereddy any ideas on this benchmarking question?

peterbell10 commented:
At a glance it looks like 1d transforms are generally slower, but nd transforms are significantly faster.
```
· `wheel_cache_size` has been renamed to `build_cache_size`. Update your `asv.conf.json` accordingly.
· Discovering benchmarks
· Running 5 total benchmarks (1 commits * 1 environments * 5 benchmarks)
[  0.00%] ·· Benchmarking existing-py_usr_bin_py
[ 10.00%] ··· Running (fft_basic.Fft.time_fft--)....
[ 50.00%] ··· Running (fft_basic.RFft.time_rfft--).
[ 60.00%] ··· fft_basic.Fft.time_fft                                          ok
[ 60.00%] ··· ====== ======= =============== ============= ============
              --                               module                  
              -------------- ------------------------------------------
               size    type   scipy.fftpack    scipy.fft    numpy.fft  
              ====== ======= =============== ============= ============
               100     real    2.74±0.06μs     5.17±0.1μs   8.97±0.2μs 
               100    cmplx     2.92±0.1μs    5.08±0.06μs   8.34±0.2μs 
               256     real    3.35±0.07μs     5.90±0.1μs   10.5±0.2μs 
               256    cmplx     3.98±0.1μs     8.10±0.7μs   9.80±0.1μs 
               512     real     5.33±0.6μs     7.59±0.2μs   14.1±0.3μs 
               512    cmplx     6.48±0.8μs     9.06±0.2μs   13.6±0.2μs 
               1000    real     7.95±0.2μs     11.9±0.1μs   21.7±0.5μs 
               1000   cmplx     10.8±0.2μs     16.5±0.7μs   20.7±0.4μs 
               1024    real     8.18±0.7μs     11.3±0.2μs   21.1±0.4μs 
               1024   cmplx     12.2±0.6μs     15.3±0.4μs   20.6±0.6μs 
               2048    real     13.6±0.4μs      20.5±3μs     39.0±1μs  
               2048   cmplx     23.0±0.3μs     30.1±0.5μs   36.5±0.4μs 
               4096    real     28.6±0.8μs     36.0±0.9μs    70.8±2μs  
               4096   cmplx      47.0±1μs       58.0±2μs    67.5±0.7μs 
               8192    real      63.7±1μs       78.4±2μs     159±7μs   
               8192   cmplx      106±2μs        130±4μs      149±3μs   
              ====== ======= =============== ============= ============

[ 70.00%] ··· fft_basic.Fft.time_ifft                                         ok
[ 70.00%] ··· ====== ======= =============== ============ ============
              --                               module                 
              -------------- -----------------------------------------
               size    type   scipy.fftpack   scipy.fft    numpy.fft  
              ====== ======= =============== ============ ============
               100     real     2.67±0.1μs    5.00±0.2μs   10.4±0.3μs 
               100    cmplx    3.56±0.07μs    4.95±0.1μs   10.4±0.3μs 
               256     real    3.52±0.06μs    6.10±0.4μs   12.5±0.2μs 
               256    cmplx     4.92±0.2μs    6.76±0.2μs    15.1±3μs  
               512     real    4.83±0.07μs    8.04±0.2μs   17.1±0.2μs 
               512    cmplx     8.24±0.1μs    9.58±0.6μs   16.6±0.5μs 
               1000    real     8.80±0.3μs    13.6±0.9μs   25.4±0.9μs 
               1000   cmplx     18.5±0.8μs    16.3±0.6μs   25.4±0.8μs 
               1024    real      9.55±1μs     11.5±0.4μs   24.9±0.7μs 
               1024   cmplx     15.7±0.5μs    16.3±0.5μs   24.7±0.4μs 
               2048    real     15.1±0.3μs    20.2±0.8μs   43.4±20μs  
               2048   cmplx     31.5±0.7μs    31.2±0.5μs    42.6±1μs  
               4096    real      33.4±1μs      40.8±1μs    77.8±0.6μs 
               4096   cmplx     76.5±40μs      59.6±1μs     79.0±2μs  
               8192    real      72.4±1μs      80.8±2μs     170±5μs   
               8192   cmplx      143±10μs      137±4μs      175±5μs   
              ====== ======= =============== ============ ============

[ 80.00%] ··· fft_basic.Fftn.time_fftn                                        ok
[ 80.00%] ··· ========== ======= =============== =============
                 size      type       module                  
              ---------- ------- --------------- -------------
               100x100     real   scipy.fftpack     231±20μs  
               100x100     real     scipy.fft       141±3μs   
               100x100     real     numpy.fft       210±20μs  
               100x100    cmplx   scipy.fftpack     222±6μs   
               100x100    cmplx     scipy.fft       179±2μs   
               100x100    cmplx     numpy.fft       203±7μs   
               1000x100    real   scipy.fftpack    2.27±0.1ms 
               1000x100    real     scipy.fft     1.00±0.06ms 
               1000x100    real     numpy.fft      4.06±0.7ms 
               1000x100   cmplx   scipy.fftpack    2.22±0.2ms 
               1000x100   cmplx     scipy.fft     1.37±0.08ms 
               1000x100   cmplx     numpy.fft     2.40±0.07ms 
               256x256     real   scipy.fftpack   1.35±0.02ms 
               256x256     real     scipy.fft       611±20μs  
               256x256     real     numpy.fft      1.54±0.6ms 
               256x256    cmplx   scipy.fftpack    1.41±0.1ms 
               256x256    cmplx     scipy.fft       879±30μs  
               256x256    cmplx     numpy.fft      2.64±0.7ms 
               512x512     real   scipy.fftpack    7.19±0.2ms 
               512x512     real     scipy.fft      2.85±0.3ms 
               512x512     real     numpy.fft      10.4±0.4ms 
               512x512    cmplx   scipy.fftpack    7.97±0.2ms 
               512x512    cmplx     scipy.fft      5.17±0.5ms 
               512x512    cmplx     numpy.fft       10.4±3ms  
              ========== ======= =============== =============

[ 90.00%] ··· fft_basic.RFft.time_irfft                                       ok
[ 90.00%] ··· ====== =============== ============= ============
              --                       module                  
              ------ ------------------------------------------
               size   scipy.fftpack    scipy.fft    numpy.fft  
              ====== =============== ============= ============
               100      3.28±0.6μs    4.97±0.06μs   12.0±0.4μs 
               256      3.92±0.7μs     6.04±0.9μs   13.5±0.9μs 
               512      4.75±0.3μs     7.81±0.6μs   16.0±0.3μs 
               1000     7.85±0.4μs     11.6±0.2μs   20.9±0.5μs 
               1024     7.02±0.2μs     10.4±0.1μs   19.6±0.6μs 
               2048     12.7±0.8μs     19.5±0.9μs   28.6±0.2μs 
               4096     31.3±0.7μs     38.6±0.7μs   51.3±0.8μs 
               8192      69.6±2μs       81.9±2μs     103±1μs   
              ====== =============== ============= ============

[100.00%] ··· fft_basic.RFft.time_rfft                                        ok
[100.00%] ··· ====== =============== ============ ============
              --                       module                 
              ------ -----------------------------------------
               size   scipy.fftpack   scipy.fft    numpy.fft  
              ====== =============== ============ ============
               100      2.73±0.1μs    5.48±0.4μs   7.87±0.2μs 
               256      3.32±0.1μs    5.54±0.2μs   8.70±0.2μs 
               512      5.62±0.6μs    8.09±0.6μs   10.9±0.2μs 
               1000     6.90±0.1μs    11.0±0.4μs   14.4±0.2μs 
               1024     7.47±0.5μs    10.5±0.5μs   14.2±0.2μs 
               2048     12.6±0.2μs    18.1±0.8μs   21.7±0.7μs 
               4096     26.7±0.8μs     32.6±1μs     38.7±1μs  
               8192      61.4±4μs      65.2±1μs     80.2±2μs  
              ====== =============== ============ ============

Running benchmarks for Scipy version 1.4.0.dev0+5dcd761 at /mnt/d/git/scipy/build/testenv/lib/python3.7/site-packages/scipy/__init__.py

```

peterbell10 commented:
One possible explanation is the lack of a cacheing for pocketfft.

mreineck commented:
> One possible explanation is the lack of a cacheing for pocketfft.

It's definitely a slowdown, but it shouldn't be that big.
Just as a rough estimate, for a 1D real FFT I measure 6.95us with numpy 1.17 and 2.72us with "naked" pypocketft.

grlee77 commented:
It may be worth adding a prime number like 2039 to the 1D FFT test cases if you want to illustrate a case where pypocketfft is likely to show a huge improvement.

rgommers commented:
> It may be worth adding a prime number like 2039 to the 1D FFT test cases if you want to illustrate a case where pypocketfft is likely to show a huge improvement.

keeping in mind that it would be good to keep the runtime with fftpack reasonable. it's not uncommon for new benchmarks to be run on old code (that's why the imports are wrapped in try-except)

peterbell10 commented:
> (that's why the imports are wrapped in try-except)
@rgommers is it safe to assume the call to the `setup` function is wrapped in try-except as well?

peterbell10 commented:
I've added 313 as a prime size that's small enough that numpy as fftpack aren't too bad:

| size | type  | scipy.fftpack | scipy.fft | numpy.fft |
|------|-------|---------------|-----------|-----------|
| 313  | real  | 71.4±0.7μs    | 29.7±2μs  | 116±6μs   |
| 313  | cmplx | 121±4μs       | 29.2±1μs  | 118±6μs   |

peterbell10 commented:
>Just as a rough estimate, for a 1D real FFT I measure 6.95us with numpy 1.17 and 2.72us with "naked" pypocketft.

@mreineck would you mind benchmarking against `scipy.fftpack`? I see even "naked" pocketfft as being in-between fftpack and numpy.fft for regular sizes. Could just be my machine though?

mreineck commented:
> @mreineck would you mind benchmarking against scipy.fftpack? I see even "naked" pocketfft as being in-between fftpack and numpy.fft for regular sizes. Could just be my machine though?

I'll try. Maybe it's also a difference in compiler flags ... which kind of flags are set when this is compiled for scipy? Just standard "-O2", I guess...

mreineck commented:
I recompiled pypocketfft with "-O2" only and ran a very basic benchmark in ipython:

```
Python 3.7.3 (default, Apr  3 2019, 05:39:12) 
Type "copyright", "credits" or "license" for more information.

IPython 5.8.0 -- An enhanced Interactive Python.
?         -> Introduction and overview of IPython's features.
%quickref -> Quick reference.
help      -> Python's own help system.
object?   -> Details about 'object', use 'object??' for extra details.

In [1]: import numpy as np

In [2]: import scipy.fftpack

In [3]: import pypocketfft as ppf

In [4]: np.version.full_version
Out[4]: '1.16.2'

In [5]: scipy.version.full_version
Out[5]: '1.1.0'

In [6]: a=np.zeros(100)+0j

In [7]: %timeit np.fft.fftn(a)
The slowest run took 28.06 times longer than the fastest. This could mean that an intermediate result is being cached.
100000 loops, best of 5: 10.9 ��s per loop

In [8]: %timeit scipy.fftpack.fftn(a)
/usr/lib/python3/dist-packages/scipy/fftpack/basic.py:160: FutureWarning: Using a non-tuple sequence for multidimensional indexing is deprecated; use `arr[tuple(seq)]` instead of `arr[seq]`. In the future this will be interpreted as an array index, `arr[np.array(seq)]`, which will result either in an error or a different result.
  z[index] = x
The slowest run took 101.44 times longer than the fastest. This could mean that an intermediate result is being cached.
100000 loops, best of 5: 11.3 ��s per loop

In [9]: %timeit ppf.fftn(a)
The slowest run took 57.16 times longer than the fastest. This could mean that an intermediate result is being cached.
100000 loops, best of 5: 3.9 ��s per loop

In [10]: 
```

Somehow this doesn't really match at all with your results, @peterbell10; not sure what's going on here.

Hardware is a fairly standard laptop with a Core i5-5300U.

peterbell10 commented:
Can you compare against `scipy.fftpack.fft` and `numpy.fft.fft` specifically since they avoid the expensive call to `_init_nd_shape_and_axes` that I had to remove.

mreineck commented:
Sure:

```
In [2]: %timeit np.fft.fft(a)
   ...: 
The slowest run took 29.89 times longer than the fastest. This could mean that an intermediate result is being cached.
100000 loops, best of 5: 7.59 ��s per loop

In [3]: %timeit scipy.fftpack.fft(a)
   ...: 
The slowest run took 32.71 times longer than the fastest. This could mean that an intermediate result is being cached.
100000 loops, best of 5: 3.74 ��s per loop
```

peterbell10 commented:
Interesting, your results are slightly closer to the fftpack results than mine. Either way, I think the main difference is probably in my python code. I'll see if I can move some of it over to the C++ side and benchmark again.

tylerjereddy commented:
> is it possible to get graphs from the output of runtests.py --bench

I know you can do `asv publish` followed by `asv preview`, but I would mostly just use that for tracking over the history of a few commits.

peterbell10 commented:
I was able to move some of the python logic over to c++ and have brought down small FFT times by ~1-2 us. 

```
In [1]: import scipy.fft; import scipy.fftpack; import numpy as np
In [2]: x = np.random.rand(100) + 1.j * np.random.rand(100)
In [3]: %timeit scipy.fft.fft(x)
4.15 µs ± 44.7 ns per loop (mean ± std. dev. of 7 runs, 100000 loops each)

In [4]: %timeit scipy.fftpack.fft(x)
2.76 µs ± 109 ns per loop (mean ± std. dev. of 7 runs, 100000 loops each)
```
Even still, `scipy.fftpack` is faster than `scipy.fft` for 1d FFTs. @mreineck's comparison is unfair in that `scipy.fftpack` has overheads from the python side as well. What if we ignore the python overheads of both and call the native extensions directly?
```
In [5]: %timeit scipy.fft._pocketfft.pypocketfft.fftn(x)
3.15 µs ± 86.8 ns per loop (mean ± std. dev. of 7 runs, 100000 loops each)

In [6]: %timeit scipy.fft._fftpack._fftpack.zfft(x)
1.12 µs ± 12.9 ns per loop (mean ± std. dev. of 7 runs, 1000000 loops each)
```
fftpack now has even more of a lead, this is definitely not my python code as it has even less overhead than fftpack's does. So, what if we can remove the cache from the equation?
```
In [7]: %timeit scipy.fft._fftpack._fftpack.destroy_zfft_cache()
135 ns ± 3.81 ns per loop (mean ± std. dev. of 7 runs, 10000000 loops each)

In [8]: %timeit scipy.fft._fftpack._fftpack.zfft(x); scipy.fft._fftpack._fftpack.destroy_zfft_cache()
4.49 µs ± 84.2 ns per loop (mean ± std. dev. of 7 runs, 100000 loops each)
```

fftpack is 4.5 us (without caching) and pocketfft is 3.2 us. Like for like, pocketfft is much faster but caching brings fftpack on top. Clealy caching is quite important here.

Profiling pocketfft shows that for n=100, there is a roughly 50-50 split between planning cost and actually executing the transform. So, assuming a cache would roughly half the time we can estimate pocketfft would take ~1.6 us, or ~2.6 us after the python overhead. Bringing it in line with cached fftpack. 

Now the question is, do we want to add a cache?

peterbell10 commented:
@mreineck 
1. Would a cache of `pocketfft_c` and `pocketfft_r` elimintate planning costs for repeated transforms?
2. Are these plans constant throughout the transform, i.e. could they be shared without locking?

mreineck commented:
> fftpack is 4.5 us (without caching) and pocketfft is 3.2 us. Like for like, pocketfft is much faster but caching brings fftpack on top. Clealy caching is quite important here.

For very short 1D transforms, caching will help, but I don't think it's important. The case of someone calling thousands of individual, short 1D FFTs is (I hope!) extremely uncommon. And replacing this by an appropriate transform along one axis of a 2D array will bring massive improvements.

Also, it's very important how the caching is done. `sclpy` currently has a caching mechanism written in C, which is very efficient but fairly inconvenient in terms of memory management.
`numpy` had caching implemented in Python, but that was extremely slow and would hurt rather than help in the situation we are discussing.

mreineck commented:
>    Would a cache of pocketfft_c and pocketfft_r elimintate planning costs for repeated transforms?

Yes. But to find a cached plan quickly enough, caching would have to be done on the C++ side.

>    Are these plans constant throughout the transform, i.e. could they be shared without locking?

Yes, they are used entirely read-only during the transform itself.

peterbell10 commented:
It's definitely true that micro-benchmarking the exact same size fft over and over it biasing it in favour in caching. However, also don't think that a small C++ cache would be very hard to implement. 

Another thing worth noting is that your plans are symmetric so the common case of `ifft(fft(x))` would get a speed up.

mreineck commented:
I'm certainly not trying to stop you :)

But I think it's quite nontrivial nonetheless.
You need a data structure that supports
- hashtable/dictionary-like lookup
- LRU removal of old plans
- locks against concurrent plan generation/destruction
- safeguards against destroying a plan which is currently in use
- some heuristics how many plans (maybe depending on plan length) should be stored.

peterbell10 commented:
>You need a data structure that supports
>- hashtable/dictionary-like lookup

Not necessarily: for small caches a simple linear search ought to be just as fast, if not faster.

>- locks against concurrent plan generation/destruction
>- safeguards against destroying a plan which is currently in use

Concurrent generation doesn't effect correctness and destruction issues are easily resolved by using `std::shared_ptr`.

larsoner commented:
> The case of someone calling thousands of individual, short 1D FFTs is (I hope!) extremely uncommon.

One application that is like this is overlap add FIR filtering, where you convolve one very long signal using one very short one using repeated small FFTs. So use cases do exist.

However, caching is something we can add later without breaking backward compatibility. I think we should open a separate issue for this where we discuss how to do it. For example we could automatically cache as has been discussed. Or we could allow optionally manually passing in a plan, which is what you might want to do in OLA filtering.

So let's focus on the minimal set of changes required to get this merged. I would say that beating fftpack and np.fft in all cases (esp. rare repeated tiny FFTs) is not a requirement, and that performance is already good enough for the first implementation to be merged. @peterbell10 if you agree, please open a new issue like you did for openmp.

@peterbell10 do you see any remaining blockers for merge? If not, @mreineck would you be able to do a review of the code? It would certainly help to have your expertise here.

peterbell10 commented:
I agree that cacheing is non-essential. I'll open an issue. I don't think there are any specific blockers. Tests are green and benchmarks are still much better than `numpy.fft` in all cases.

@mreineck if you review the code, would you mind looking at `scipy_features.h`. Particularly the `xfftn_real` implementation.  I would be interested to know if you think the repeated mirroring could be done in a simpler way.

mreineck commented:
> One application that is like this is overlap add FIR filtering, where you convolve one very long signal using one very short one using repeated small FFTs. So use cases do exist.

Very good point, thanks for mentioning this!
(Although for best performance I'd still put all these short overlapping array snippets into a 2D array and then FFT one axis. Then vectorization kicks in and will beat the 1D approach. But that's maybe too much effort.)

> @mreineck would you be able to do a review of the code? It would certainly help to have your expertise here.

Sure! It might take a few days though; I'll be quite busy till next Wednesday or so.

peterbell10 commented:
Travis failure is in linprog:
`FAILED ../../optimize/tests/test_linprog.py::TestLinprogSimplexBland::test_unbounded_below_no_presolve_corrected`

mdhaber commented:
That's weird. I restarted the build to see if it happens again.

larsoner commented:
> It might take a few days though; I'll be quite busy till next Wednesday or so.

That's a fine timeline, @grlee77 and I (and anyone else interested) can hopefully take a closer look before then too. Then after your feedback we'll hopefully be good to merge.

@peterbell10 I'm the meantime I'm sure there are also more generic backend related things to continue exploring.

grlee77 commented:
Yes, I should be able to take a closer look this weekend.

peterbell10 commented:
@mdhaber looks like it's a deprecation warning:

>DeprecationWarning: Calling nonzero on 0d arrays is deprecated, as it behaves surprisingly. Use `atleast_1d(cond).nonzero()` if the old behavior was intended. If the context of this warning is of the form `arr[nonzero(cond)]`, just use `arr[cond]`.

mdhaber commented:
Thanks. I'll make an issue.

jwd0023 commented:
For anyone following, @mdhaber is working on the fix. 

Issue: #10303 
Fix:     #10306

grlee77 commented:
It doesn't show up in the diffs for the PR, but some text at the top of the FFTPack tests are outdated and will not work. For example:
https://github.com/scipy/scipy/blob/498c67673b39aa89e82a81f2c5f7e9a92d4b6ba5/scipy/fft/_fftpack/tests/test_basic.py#L9
The same outdated text is also present in a few other files in that same folder.

It seems that the new usage would be:
`python -c 'import scipy.fft;scipy.fft.test()'`
(or more specifically `python -c 'import scipy.fft;scipy.fft._fftpack.test()'`)

Similarly, one should not try the following to run FFTPack tests:
`python runtests.py -s fftpack`
because the tests are no longer in that folder and will not be found.

Instead one has to do:
```python runtests.py -s fft``` or ```python runtests.py -s fft._fftpack```

larsoner commented:
> If the intention is to move scipy.fftpack over to using pocketfft as well, then it creates a circular dependency issue. fft depends on fftpack for the cosine and sine transforms; fftpack would depend on fft for the normal FFTs.

A simpler way to solve this is to nest the imports, e.g. in `scipy/fftpack`:
```
def fft(...)
    from ..fft import fft
    returnn fft(...)
```
This might be a simpler approach in the end, especially in light of what @grlee77 points out in terms of new problems we're likely encountering now.

> In the current state (where fftpack doesn't use pocketfft) it's not necessary and I could revert it.

Let's do this for now and deal with getting `scipy.fftpack` to use `scipy.fft` later if we want it. I'm not sure it will be necessary.

grlee77 commented on scipy/fft/_pocketfft/setup.py:
```diff
@@ -0,0 +1,76 @@
+
+        import sys
+        if sys.platform == 'darwin':
+            args += ['-stdlib=libc++', '-mmacosx-version-min=10.7']
```
I would also use `has_flag` for `-stdlib=libc++` (some OS X users may want to build with gcc rather than clang and I don't think that will work with that flag).

 Also, I think conda-forge and anaconda currently use 10.9 for `MACOSX_DEPLOYMENT_TARGET`. Setting `-mmacosx-version-min=10.7` will override this. Is this okay to do as long as the version here is older or does setting this require the user to have an older SDK installed?

peterbell10 replied:
From reading around, it seems that the flag mainly adds compiler warnings when using unavailable APIs. It doesn't sound like it changes which version of the SDK is linked.

peterbell10 commented:
I have restored fftpack back to it's original location.

mreineck commented:
> @mreineck if you review the code, would you mind looking at scipy_features.h. Particularly the xfftn_real implementation. I would be interested to know if you think the repeated mirroring could be done in a simpler way.

I think it will be hard to find a simpler way. But if I read the code right, you are performing one mirroring step per dimension instead of doing everything in one go, which is probably suboptimal.
Doing all the mirroring at once requires fairly ugly index calculations. I have implemented an iterator which can do this; you can find it on the pypocketfft `master` branch, where it is used in `complex2hartley`.

On the `new_norm` branch I've gone a bit further and written an xfftn_real equivalent based on this iterator. This branch also eliminates the need for passing normalization factors; it uses an integer `inorm` argument instead, where 0 means "no normalization", 1 means "divide by sqrt(size)" and 2 means "divide by size".

Not sure whether it is still an option for you to switch to this code at the current stage, however ... sorry for implementing all of this only now, but it's hard to anticipate what will be needed most urgently.

larsoner commented:
>> I would be interested to know if you think the repeated mirroring could be done in a simpler way.

> But if I read the code right, you are performing one mirroring step per dimension instead of doing everything in one go, which is probably suboptimal. Doing all the mirroring at once requires fairly ugly index calculations. I have implemented an iterator which can do this; you can find it on the pypocketfft master branch, where it is used in complex2hartley.

Is this the sort of thing that improves performance by a factor of 2x, or more like 1.02x for typical use cases? If the latter I'd rather leave it for a subsequent PR unless it's simple/easy to do it already, but based on what you say it sounds non-trivial. Making code more complex for the sake of speed in this PR will make it harder to find bugs / ensure we're testing properly and harder to review the code.

mreineck commented:
> Is this the sort of thing that improves performance by a factor of 2x, or more like 1.02x for typical use cases?

The reordering part will be faster by integral factors for multi-D transforms, but I cannot really say what the overall speedup (including the FFT cost) will be.

> Making code more complex for the sake of speed in this PR will make it harder to find bugs / ensure we're testing properly and harder to review the code.

I agree, but the added complexity is very well localized in the new `rev_iter` class and will be re-used in multiple places. Also it's fairly easy to test for correctness.

larsoner commented:
Okay I'll let @peterbell10 decide if it's worth it after working. We would probably want to do add benchmark that shows there is indeed a measurable benefit for some FFT configuration at least. And ensure it is tested for correctness.

Let's just keep in mind that we don't have to have things fully optimized in this first PR. Once a working version is in master, it's also easier for others to add little (or big) optimizations. This is another reason why I'm pushing for incremental progress in master rather than additional progress in this PR.

mreineck commented:
That's all perfectly fine!

I mainly wanted to inform everyone that functionality very similar to what @peterbell10 did in `scipy_features.h` will be available in vanilla pypocketfft in the future, to avoid potential redundant work.

peterbell10 commented:
>if I read the code right, you are performing one mirroring step per dimension instead of doing everything in one go, which is probably suboptimal.

If I'm not mistaken, doing the mirrorings simultaneously would lead to a heavily strided memory access pattern. Considering the effects of caching and pre-fetching, it's not immediately obvious to me that it would actually be faster. So, I took for the simpler route.

>[the `new_norm` branch] uses an integer `inorm` argument instead, where 0 means "no normalization", 1 means "divide by sqrt(size)" and 2 means "divide by size".

It's nice to see that the two versions are converging, but why integer arguments instead of an `enum`? Much more readable imo.

>Not sure whether it is still an option for you to switch to this code at the current stage

At this stage, I'm in favour of updating in a seperate PR. This one is quite large already.

peterbell10 commented:
>I mainly wanted to inform everyone that functionality very similar to what @peterbell10 did in scipy_features.h will be available in vanilla pypocketfft in the future, to avoid potential redundant work.

@mreineck to avoid future divergence, is there any way to colaborate on (py)pocketfft itself? Perhaps a github mirror of the repo?

mreineck commented:
> If I'm not mistaken, doing the mirrorings simultaneously would lead to a heavily strided memory access pattern.

Not necessarily, we just need to make sure that the innermost index is the only with the smallest stride. Then the forward iterator will advance linearly, while the "backward" iterator will go mostly linearly backward , with the exception of one "jump" per line. Overall it should be better than touching the memory several times.

> It's nice to see that the two versions are converging, but why integer arguments instead of an enum? Much more readable imo.

Pure pragmatism. If the forward transform is scaled by `inorm`, the backward one must be scaled with `2-inorm` :)

> At this stage, I'm in favour of updating in a seperate PR. This one is quite large already.

I'm absolutely happy with that.

> @mreineck to avoid future divergence, is there any way to colaborate on (py)pocketfft itself? Perhaps a github mirror of the repo?

That's possible, but then I'll have to juggle with two upstreams. I can also provide external access to the Gitlab server; all I need is an email address.

mreineck commented:
> If the forward transform is scaled by inorm, the backward one must be scaled with 2-inorm :)

Alternatively, `scaling_factor := sqrt(N)**(-inorm)`

I'm a bit skeptical about all the `enum` conversion magic going on at the language boundary.

peterbell10 commented:
>>If I'm not mistaken, doing the mirrorings simultaneously would lead to a heavily strided memory access pattern.

>Not necessarily, [...] the "backward" iterator will go mostly linearly backward , with the exception of one "jump" per line. Overall it should be better than touching the memory several times.

Yes, the access is still linear but the stides mean you're not writing to the full cache lines and those "jumps" happen much more frequently. Personally, I would want to measure the difference.

mreineck commented on scipy/fft/_helper.py:
```diff
@@ -0,0 +1,52 @@
+    >>> b = fft.fft(a)
+
+    Zero-padding to the next regular length reduces computation time to
+    1.3 ms, a speedup of 33 times:
```
That seems to be a typo, should be "13 times"

peterbell10 replied:
Thanks, fixed

mreineck commented on scipy/fft/_pocketfft/scipy_features.h:
```diff
@@ -0,0 +1,233 @@
+
+  // Convert strides from bytes to objects
+  for (auto & s : strides_out)
+    s /= sizeof(std::complex<T>);
```
I'm not sure that this will work in all situations. It's not unthinkable that the stride is not cleany divisible by sizeof(std::complex<T>) (e.g. in structured arrays).

mreineck replied:
Nevermind ... I realized that you are computing the strides for an array that you just created yourself; that should always be safe. Problems will start once `inplace` or `out` arguments are used, since you no longer have control over the strides then.

grlee77 commented:
I can confirm that compilation now works on OS X with gcc as well (I tried on OSX 10.10 laptop with gcc 5.4).

Also, after moving the fftpack files back, testing via `scipy.fftpack.test()` works again as in the past as well.

larsoner commented:
I should be able to give this a full review tomorrow or Friday.

In the meantime, @grlee77 @mreineck are you happy with the latest changes? If so feel free to "approve" the PR in reviewing mode. It would be good to make sure everyone is on board since this is an important starting point.

grlee77 commented:
Yes, I am happy with it (with the caveat that I did not review the new code in `scipy_features.h` very closely, although I think @mreineck has just recently reviewed that)

larsoner commented on benchmarks/benchmarks/fft_basic.py:
```diff
@@ -53,47 +54,43 @@ def setup(self, size, cmplx, module):
         else:
             self.x = random([size]).astype(double)
 
+        self.fft = eval(module + '.fft')
```
Maybe `getattr(module, 'fft')` instead of `eval` (which is a bit overkill)?

larsoner commented on scipy/fft/_helper.py:
```diff
@@ -0,0 +1,52 @@
+import scipy.fftpack as _fftpack
+
```
PEP8 two newlines before function def

larsoner commented on scipy/fft/_helper.py:
```diff
@@ -0,0 +1,52 @@
+    """
+
+    # TODO: Complex transforms are also fast for 11-smooth lengths
+    return _fftpack.next_fast_len(target)
```
Is it possible that the `next_fast_len` for `fftpack` and `pocketfft` are different? Do they have the same radix-es?

peterbell10 replied:
For real transforms pocketfft only has 2, 3 & 5 which is the same as fftpack. For complex to complex transforms there are also radices 7 & 11 as per the comment.

larsoner commented on scipy/fft/_pocketfft/tests/test_basic.py:
```diff
@@ -0,0 +1,1004 @@
+
+    def test_size_accuracy(self):
+        # Sanity check for the accuracy for prime and non-prime sized inputs
+        for size in LARGE_COMPOSITE_SIZES + LARGE_PRIME_SIZES:
```
It would be good to have a test somewhere that for the composite sizes, `next_fast_len(x) == x`. Not 100% relevant to your changes here, but might eventually relate to the radix question above, and ensure that you are actually testing things that are composite sizes.

peterbell10 replied:
Would you put this test in the pocketfft tests or in the fft module tests?

larsoner replied:
Probably fft, but either is fine

peterbell10 replied:
I've added a test.

larsoner commented on scipy/fft/_pocketfft/version.md:
```diff
@@ -0,0 +1,11 @@
+pypocketfft version
+-------------------
+
+SciPy currently vendors [pypocketfft][repo] at commit hash:
```
it would be good to add the date of the commit here, too, so no investigation is needed to tell how far out of date we are time-wise

peterbell10 replied:
I've added the full git log entry

mreineck commented on scipy/fft/_pocketfft/scipy_features.h:
```diff
@@ -0,0 +1,233 @@
+  {
+  if (norm == norm_t::none) return T(1);
+
+  T N(1);
```
Sorry for noticing this so late ...

This snippet is actually problematic for single precision and large multi-D shapes, because the product of the axis lengths may no longer be representable exactly in `float`.
Also, doing the factor computation in the type of the transform introduces more rounding errors than strictly necessary.
How about this:
```
size_t N(1);
  for (auto a : axes)
    N *= shape[a];

  if (norm == norm_t::size) return T(1.L / N);
  if (norm == norm_t::ortho) return T(1.L / sqrt(1.L*N));
```

This may be slight overkill, but it doesn't really cost anything

peterbell10 replied:
Thanks, fixed.

larsoner commented:
Okay the FFT people appear to be happy and no other complaints, in it goes. Thanks @peterbell10 !

WarrenWeckesser commented:
>  This adds a build time dependency on pybind11 and I'm not sure if it is required at run-time as well.

This should be documented in INSTALL.rst.txt.

endolith commented on scipy/fft/_pocketfft/basic.py:
```diff
@@ -0,0 +1,262 @@
+
+def rfft2(x, shape=None, axes=(-2,-1), norm=None, overwrite_x=False):
+    """
+    2-D dicsrete Fourier transform of a real sequence
```
"discrete" typo

endolith commented on scipy/fft/_pocketfft/basic.py:
```diff
@@ -0,0 +1,262 @@
+
+def irfft2(x, shape=None, axes=(-2,-1), norm=None, overwrite_x=False):
+    """
+    2-D dicsrete inverse Fourier transform of a real sequence
```
"discrete" typo

endolith commented on scipy/fft/_pocketfft/basic.py:
```diff
@@ -0,0 +1,262 @@
+    return pfft.ifftn(tmp, axes, norm, overwrite_x, _default_workers)
+
+def rfftn(x, shape=None, axes=None, norm=None, overwrite_x=False):
+    """Return multi-dimentional discrete Fourier transform of real input"""
```
"multi-dimentional" typo

endolith commented on scipy/fft/_pocketfft/basic.py:
```diff
@@ -0,0 +1,262 @@
+
+def rfft(x, n=None, axis=-1, norm=None, overwrite_x=False):
+    """
+    Discrete Fourier transform of a real sequence.
```
I guess these docstrings will be expanded to the same level of details the old ones in a different PR?

peterbell10 replied:
The user-facing API is in `scipy/fft/basic.py` which has the full docstrings.

peterbell10 replied:
See the [rendered docs](https://scipy.github.io/devdocs/generated/scipy.fft.fft.html#scipy.fft.fft)

peterbell10 commented:
I've added a release note to the wiki page:
https://github.com/scipy/scipy/wiki/Release-note-entries-for-SciPy-1.4.0#scipyfft-added

rgommers commented:
That looks great, thanks Peter

WarrenWeckesser commented:
I was about to update some old code to use `scipy.fft` instead of `scipy.fftpack`, but I see that `scipy.fft` doesn't have a replacement for [`scipy.fftpack.diff`](https://scipy.github.io/devdocs/reference/generated/scipy.fftpack.diff.html).  Was there ever a plan to add `diff` (or something similar) to `scipy.fft`?  Or is something equivalent already available somewhere else?

mreineck commented:
I vaguely remember that we discussed that at the time and think we agreed that no replacement was needed, since `diff` was not documented and not used. That said, it should be easy to add as a post-processing step, correct?

WarrenWeckesser commented:
The docstring that I linked to is not complete, but I wouldn't say the function was not documented.  It is definitely part of the public API.   It might not be widely used, but:

* I used it years ago in the SciPy cookbook example ["Korteweg de Vries equation"](https://scipy.github.io/devdocs/reference/generated/scipy.fftpack.diff.html), and at least two people have used that example as a model for their own code:
  * [How to implement boundary value in Python using spectral methods](https://math.stackexchange.com/questions/3146228/how-to-implement-boundary-value-in-python-using-spectral-methods)
  * [how to use solve_ivp solve Partial Differential Equations with spectral method?](https://stackoverflow.com/questions/60974226/how-to-use-solve-ivp-solve-partial-differential-equations-with-spectral-method)

  Presumably others have used the reference.
* It shows up now and then on stackoverflow, e.g.
  * [`numpy.diff` and `scipy.fftpack.diff` giving different results when differentiating](https://stackoverflow.com/questions/42197460/numpy-diff-and-scipy-fftpack-diff-giving-different-results-when-differentiat)
  * [Computing numeric derivative via FFT - SciPy](https://stackoverflow.com/questions/62102774/computing-numeric-derivative-via-fft-scipy)
  * [Derivative of an array in python?](https://stackoverflow.com/questions/20044096/derivative-of-an-array-in-python)
* Some other random mentions on the web:
  * [Practical notes on selected numerical methods with examples (PDF)](https://www.ioc.ee/~dima/publications/r312_15.pdf)
  * ["SciPy for Signal Processing"](https://hub.packtpub.com/scipy-signal-processing/)

Having said all that, apparently this is the first time someone has asked about the "missing" functions, so I guess the demand isn't really that high.  Certainly the *method* is widely used, but I suspect the existence of the function is not well known.  If you understand the method, it is not hard to implement, but I wouldn't call our current implementation trivial:  https://github.com/scipy/scipy/blob/39fa4852bdefcd9675dc65e4f4ad059356cee265/scipy/fftpack/_pseudo_diffs.py#L49-L74
